### PR TITLE
[bcr-wdc-key-service] Fix mint verification

### LINKED ISSUE #144: Mint will sign blinds for more than discounted offer
The mint will sign blinded messages for more than the discounted offer made.

```
e2e-tests-1  | 2025-05-09T11:42:45.649081Z  INFO bcr_wdc_e2e_tests: START EBILL MINTING TEST
e2e-tests-1  | 2025-05-09T11:42:45.735105Z  INFO bcr_wdc_e2e_tests: Bill created bill_amount=91811 bill_id="6fPjW9JcYD6VEwPQsLxNAzx9BngKzWUbs4ryoY1YDHsF"
e2e-tests-1  | 2025-05-09T11:42:45.735135Z  INFO bcr_wdc_e2e_tests: Requesting to mint the bill
e2e-tests-1  | 2025-05-09T11:42:45.783320Z  INFO bcr_wdc_e2e_tests: Mint Request Accepted, waiting for admin to process quote_id=8073586f-1fd8-42f9-8c59-874e34d2dda3
e2e-tests-1  | 2025-05-09T11:42:45.783338Z  INFO bcr_wdc_e2e_tests: Admin sending discounted offer discounted=90892
e2e-tests-1  | 2025-05-09T11:42:45.863621Z  INFO bcr_wdc_e2e_tests: Quote is offered amount=0.00090892 BTC ttl=2026-05-09 11:42:45.783337676 UTC
e2e-tests-1  | 2025-05-09T11:42:45.870158Z  INFO bcr_wdc_e2e_tests: Getting mint quote status from
e2e-tests-1  | 2025-05-09T11:42:45.870167Z  INFO bcr_wdc_e2e_tests: Quote is accepted keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:45.870169Z  INFO bcr_wdc_e2e_tests: Activating keyset for quote_id: 8073586f-1fd8-42f9-8c59-874e34d2dda3
e2e-tests-1  | 2025-05-09T11:42:45.985488Z  INFO bcr_wdc_e2e_tests: Confirmed active keyset keyset_info_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:45.986154Z  INFO bcr_wdc_e2e_tests: Signing NUT20 mint request
e2e-tests-1  | 2025-05-09T11:42:45.986220Z  INFO bcr_wdc_e2e_tests: Sending NUT20 mint request
e2e-tests-1  | 2025-05-09T11:42:46.067661Z  INFO bcr_wdc_e2e_tests: Mint Successful obtained signatures amount=91811
e2e-tests-1  | 2025-05-09T11:42:46.067675Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(02f446d4d464750b0cd7a9b05e251268907a39fd21cb72f15e6b3ffb7f7fa07483) amount=Amount(1) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067682Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(034337b680966de83b251e6ab841f044d5d7835d9df95d4620a2aaf353858ade01) amount=Amount(2) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067685Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(03cf9a7caac2e94f6cf3a9903d27c5e6d597c6caed132d1b6f08ea0fc1d726b113) amount=Amount(32) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067687Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(03c6f1e0b0266503481e4543b969a0212b7d184b0b791ca8adcabf7a42dbe4b7f1) amount=Amount(128) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067690Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(02cd1b0dafe907f0cac3415a9136ffa202e2efffb598f65759fa344d210b4a4524) amount=Amount(512) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067692Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(029eaeda6ca13994d67a97b37fdcd7e8bd224182174f5e88be43f58828591d904a) amount=Amount(1024) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067697Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(032db8bc675ec23043946334ea5109edecd7f881c554b4ece9c2c25bf488e8f33a) amount=Amount(8192) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067700Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(023d047b38a72920592751364afb946e79978200351bec748bd1487caa9e2edf33) amount=Amount(16384) keyset_id=004c105fd2d5988f
e2e-tests-1  | 2025-05-09T11:42:46.067702Z  INFO bcr_wdc_e2e_tests: Signature c_=PublicKey(0372541fd20912f68b97a2dc9946c3b6ca9a11e13514089c9a6642079534e19e37) amount=Amount(65536) keyset_id=004c105fd2d5988f
```

### PR DESCRIPTION
**Pull request by stefanbitcr**
### **User description**
Please review carefully as there are big logic changes, and I'm not 100% sure what the implications are for only storing (only) the discounted amount.

When generating the keyset, I've switched to using the discounted amount.
Something to think about is potentially storing the original bill amount, as well as the discounted amount.

The discounted amount is needed to verify the exact sum of the blinds matches it.

Previously, we could mint up the total amount of the Bill, more than the discount the wildcat mint offered the user.

This PR also enforces the requirement for the amount of blinds to match exactly the discounted amount, as the wallet client is able to do so using the information provided, freeing the mint from selecting blinds.

Closes #144

___

### **PR Type**
Bug fix, Tests

___

### **Description**
- Enforces minting only up to the discounted offer amount.
  - Validates sum of blinded outputs matches discounted amount.
  - Prevents minting for more than the discounted value.

- Removes and refactors blind selection logic for minting.

- Adds comprehensive tests for minting logic and edge cases.

- Fixes in-memory persistence to correctly mark keysets as minted.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>service.rs</strong><dd><code>Enforce discounted minting and refactor mint logic with tests</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/service.rs

<li>Validates that the sum of blinded outputs equals the discounted <br>amount.<br> <li> Removes <code>select_blinds_to_target</code> and related logic.<br> <li> Refactors minting to sign all provided outputs if valid.<br> <li> Adds new async tests for minting success and failure cases.

</details>

  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/149/files#diff-0bf63ca2bfa682f3961784061646602648f6450798339852bae9daf95e784e52">+103/-186</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>inmemory.rs</strong><dd><code>Fix in-memory keyset minting state update logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-service/src/persistence/inmemory.rs

<li>Fixes <code>mark_as_minted</code> to return early after marking as minted.<br> <li> Ensures correct error handling for unknown keysets.

</details>

  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/149/files#diff-3f12dfeeadb01ce475140bd0f334bac22948a837c7d6f5ce31a4120e1a8ef560">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>service.rs</strong><dd><code>Use discounted amount for keyset generation in quotes</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/service.rs

<li>Changes keyset generation to use discounted amount instead of bill <br>sum.<br> <li> Ensures minting is based on the correct discounted value.

</details>

  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/149/files#diff-186f7e8fe93bbc7070dc420fa5c734f262aa91e2cc4a5585eecc838222112f36">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: crates/bcr-wdc-key-service/src/service.rs [hunk 1]
```diff
@@ -214,10 +214,13 @@ where
         if is_minted {
             return Err(Error::InvalidMintRequest);
         }
+        let blind_sum = outputs.iter().map(|o| u64::from(o.amount)).sum::<u64>();
+        if blind_sum != u64::from(target) {
+            return Err(Error::InvalidMintRequest);
```
Comment (stefanbitcr) [inline crates/bcr-wdc-key-service/src/service.rs:219]: No need to handle cases where the wallet client provided incorrect blind amounts

File: crates/bcr-wdc-quote-service/src/service.rs [hunk 2]
```diff
@@ -270,7 +270,7 @@ where
         let maturity_date = quote.bill.maturity_date;
         let kid = self
             .keys_hndlr
-            .generate(qid, quote.bill.sum, public_key, maturity_date)
```
Comment (stefanbitcr) [inline crates/bcr-wdc-quote-service/src/service.rs:273]: This is used to check how much we can mint later

File: crates/bcr-wdc-key-service/src/service.rs [hunk 3]
```diff
@@ -226,196 +229,110 @@ where
...
+            let blind = bcr_wdc_utils::keys::test_utils::generate_blind(keyset_id, *amount);
+            blinds.push(blind);
         }
+        blinds
     }
-    blinds
-}
 
-#[cfg(test)]
-pub mod tests {
```
Comment (stefanbitcr) [inline crates/bcr-wdc-key-service/src/service.rs:251]: The tests here don't capture the scenario of minting more than the discount, this would require an end to end test

**Review by mtbitcr — APPROVED** [review]
No merge without @codingpeanut157 approves.

**Review by codingpeanut157 — APPROVED** [review]
Very good catch!

Side note:
a potential performance improvement, very cheap to implement, would be to create a middle layer that goes in-between the Service and the KeysRepository that caches the keys locally.
Something as simple as 
```rust
struct KeysCache<KeysRepo> {
   cache: Arc<RWLock<[cashu::MintKeys, CACHE_SIZE]>>,
   repo: KeysRepo,
}

impl<KeysRepo> KeysRepository for KeysCache<KeysRepo>
where KeysRepo: KeysRepository 
{
   ...
}
```
where we cache the last `CACHE_SIZE` used sets of keys.
MintKeys never change, so we won't have to worry about cache invalidation.

We could implement the same thing in the key-client as well.
